<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>DeepScanScape: Deep Mutational Landscape Analysis • DeepScanScape</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="DeepScanScape: Deep Mutational Landscape Analysis">
<meta property="og:description" content="deepscanscape">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">DeepScanScape</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9042</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/deepscanscape.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/allydunham/deepscanscape/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="deepscanscape_files/header-attrs-2.8/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>DeepScanScape: Deep Mutational Landscape Analysis</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/allydunham/deepscanscape/blob/master/vignettes/deepscanscape.Rmd"><code>vignettes/deepscanscape.Rmd</code></a></small>
      <div class="hidden name"><code>deepscanscape.Rmd</code></div>

    </div>

    
    
<p>This vignette demonstrates usage of the DeepScanScape package by importing several deep mutational scans and analysing them. First we load the DeepScanScape package and initialise a list to hold the parsed datasets.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/allydunham/deepscanscape/">deepscanscape</a></span><span class="op">)</span>
<span class="va">dms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div id="loading-deep-mutational-scan-data" class="section level1">
<h1 class="hasAnchor">
<a href="#loading-deep-mutational-scan-data" class="anchor"></a>Loading Deep Mutational Scan Data</h1>
<p>We will use the following studies as examples:</p>
<ul>
<li>p53 (<a href="https://doi.org/10.1016/j.molcel.2018.06.012">Kotler et al. 2018</a>)</li>
<li>UBE2i (<a href="https://doi.org/10.15252/msb.20177908">Weile et al. 2017</a>)</li>
<li>Influenza Nucleoprotein (<a href="https://doi.org/10.1093/molbev/msv167">Doud et al. 2015</a>)</li>
<li>hYAP65 WW domain (<a href="https://doi.org/10.1073/pnas.1209751109">Araya et al. 2012</a>)</li>
</ul>
<p>The latter 3 are part of the Deep Landscape dataset generated by <a href="https://doi.org/10.1101/2020.05.26.116756">Dunham &amp; Beltrao (2020)</a>, while the first is a new study not used to generate the landscape. In each case we first load the data into a base R data frame and examine it’s structure, then process it so that it can be loaded into the standard format for DeepScanScape - the deep_mutational_scan object. These store deep scan data in a standard format with variant effect scores on a standard normalised scale.</p>
<div id="p53" class="section level2">
<h2 class="hasAnchor">
<a href="#p53" class="anchor"></a>p53</h2>
<p>This data comes from <a href="https://www.mavedb.org">MaveDB</a>, meaning it is in a standardised format that DeepScanScape recognises and can directly convert into a deep_mutational_scan.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df_p53</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"urn_mavedb_00000059_a_1_scores.csv"</span>, package <span class="op">=</span> <span class="st">"deepscanscape"</span><span class="op">)</span>, skip <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">df_p53</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>
<span class="co">#&gt;                     accession hgvs_nt    hgvs_pro       score</span>
<span class="co">#&gt; 1   urn:mavedb:00000059-a-1#1      NA  p.Lys190fs  0.51437850</span>
<span class="co">#&gt; 2   urn:mavedb:00000059-a-1#2      NA  p.Lys190fs  1.12150651</span>
<span class="co">#&gt; 3   urn:mavedb:00000059-a-1#3      NA  p.Lys190fs  0.12167722</span>
<span class="co">#&gt; 4   urn:mavedb:00000059-a-1#4      NA  p.Lys190fs  0.93340046</span>
<span class="co">#&gt; 5   urn:mavedb:00000059-a-1#5      NA  p.Lys190fs  0.59828604</span>
<span class="co">#&gt; 6   urn:mavedb:00000059-a-1#6      NA  p.Lys190fs  0.58229992</span>
<span class="co">#&gt; 7   urn:mavedb:00000059-a-1#7      NA  p.Lys190fs  0.46064492</span>
<span class="co">#&gt; 8   urn:mavedb:00000059-a-1#8      NA  p.Lys190fs -0.09049759</span>
<span class="co">#&gt; 9   urn:mavedb:00000059-a-1#9      NA  p.Lys190fs  0.53761134</span>
<span class="co">#&gt; 10 urn:mavedb:00000059-a-1#10      NA p.Lys190Trp -2.40583136</span></code></pre></div>
<p>In this case there are duplicate scores and multiply mutated sequences so we use the “duplicates” and “average_multi” arguments to handle them. The scores are already in log<sub>2</sub> form, so do not need any further transforming.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dms</span><span class="op">$</span><span class="va">p53</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/deep_mutational_scan.html">deep_mutational_scan</a></span><span class="op">(</span><span class="va">df_p53</span>, name <span class="op">=</span> <span class="st">"p53"</span>, gene <span class="op">=</span> <span class="st">'p53'</span>, study <span class="op">=</span> <span class="st">'Kotler et al. (2018)'</span>,
                                source <span class="op">=</span> <span class="st">"https://www.mavedb.org/scoreset/urn:mavedb:00000059-a-1/"</span>,
                                scheme <span class="op">=</span> <span class="st">"mavedb"</span>, trans <span class="op">=</span> <span class="cn">NULL</span>, na_value <span class="op">=</span> <span class="st">'impute'</span>, duplicates <span class="op">=</span> <span class="st">"mean"</span>,
                                average_multi <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">dms</span><span class="op">$</span><span class="va">p53</span>
<span class="co">#&gt; # Deep mutational scanning data</span>
<span class="co">#&gt; # Name: p53</span>
<span class="co">#&gt; # Study: Kotler et al. (2018)</span>
<span class="co">#&gt; # Gene: p53</span>
<span class="co">#&gt; # 191 positions</span>
<span class="co">#&gt; # Positional data:</span>
<span class="co">#&gt;    position wt          A       C       D       E       F       G       H</span>
<span class="co">#&gt;       &lt;dbl&gt; &lt;chr&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;</span>
<span class="co">#&gt;  1        1 T     -0.767  -0.683  -0.658  -0.239  -0.690  -0.957  -0.693 </span>
<span class="co">#&gt;  2        2 Y     -0.437  -0.717  -0.873  -0.714  -0.703  -0.269  -0.900 </span>
<span class="co">#&gt;  3        3 Q     -0.690  -0.0939 -0.884  -0.430  -0.124  -0.936  -0.617 </span>
<span class="co">#&gt;  4        4 G     -0.238  -0.0746 -0.539  -0.0621  0.0467  0       0.0259</span>
<span class="co">#&gt;  5        5 S     -0.772  -1.03   -1.00   -0.157  -0.666  -0.919  -0.997 </span>
<span class="co">#&gt;  6        6 Y     -0.307  -0.678   0.0557  0.169  -0.836   0.0550 -0.416 </span>
<span class="co">#&gt;  7        7 G     -0.540  -0.595  -0.729  -0.587  -0.675   0      -0.835 </span>
<span class="co">#&gt;  8        8 F      0.153  -0.102   0.137  -0.613   0       0.128   0.216 </span>
<span class="co">#&gt;  9        9 R     -0.722  -0.560  -0.0741 -0.320  -0.328  -0.511  -0.545 </span>
<span class="co">#&gt; 10       10 L      0.0169 -0.166  -0.576  -0.0247  0.249   0.270   0.120 </span>
<span class="co">#&gt; # … with 181 more rows, and 34 more variables: I &lt;dbl&gt;, K &lt;dbl&gt;, L &lt;dbl&gt;,</span>
<span class="co">#&gt; #   M &lt;dbl&gt;, N &lt;dbl&gt;, P &lt;dbl&gt;, Q &lt;dbl&gt;, R &lt;dbl&gt;, S &lt;dbl&gt;, T &lt;dbl&gt;, V &lt;dbl&gt;,</span>
<span class="co">#&gt; #   W &lt;dbl&gt;, Y &lt;dbl&gt;, name &lt;chr&gt;, impute_A &lt;dbl&gt;, impute_C &lt;dbl&gt;,</span>
<span class="co">#&gt; #   impute_D &lt;dbl&gt;, impute_E &lt;dbl&gt;, impute_F &lt;dbl&gt;, impute_G &lt;dbl&gt;,</span>
<span class="co">#&gt; #   impute_H &lt;dbl&gt;, impute_I &lt;dbl&gt;, impute_K &lt;dbl&gt;, impute_L &lt;dbl&gt;,</span>
<span class="co">#&gt; #   impute_M &lt;dbl&gt;, impute_N &lt;dbl&gt;, impute_P &lt;dbl&gt;, impute_Q &lt;dbl&gt;,</span>
<span class="co">#&gt; #   impute_R &lt;dbl&gt;, impute_S &lt;dbl&gt;, impute_T &lt;dbl&gt;, impute_V &lt;dbl&gt;,</span>
<span class="co">#&gt; #   impute_W &lt;dbl&gt;, impute_Y &lt;dbl&gt;</span></code></pre></div>
</div>
<div id="ube2i" class="section level2">
<h2 class="hasAnchor">
<a href="#ube2i" class="anchor"></a>UBE2i</h2>
<p>This is another dataset from MaveDB, so can be processed in the same way. However, in this case the data was processed using the VAMP-Seq method (<a href="https://doi.org/10.1038/s41588-018-0122-z">Matreyek et al. 2018</a>) and scores are on a different scale, with neutral variants <span class="math inline">\(\approx 1\)</span> and null variants <span class="math inline">\(\approx 0\)</span>. DeepScanScape requires scores to be transformed onto the <span class="math inline">\(log_2ER\)</span> scale, including built in transformations for common score formats such as VAMP-Seq. More unusual scores require a custom function to transform them.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df_ube2i</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"urn_mavedb_00000001-a-1_scores.csv"</span>, package <span class="op">=</span> <span class="st">"deepscanscape"</span><span class="op">)</span>, skip <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>
<span class="va">dms</span><span class="op">$</span><span class="va">ube2i</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/deep_mutational_scan.html">deep_mutational_scan</a></span><span class="op">(</span><span class="va">df_ube2i</span>, name <span class="op">=</span> <span class="st">"UBE2i"</span>, gene <span class="op">=</span> <span class="st">'UBE2i'</span>, study <span class="op">=</span> <span class="st">'Weile et al. (2017)'</span>,
                                  scheme <span class="op">=</span> <span class="st">"mavedb"</span>, trans <span class="op">=</span> <span class="st">"vamp-seq"</span>, na_value <span class="op">=</span> <span class="st">'impute'</span>, duplicates <span class="op">=</span> <span class="st">"mean"</span>,
                                  average_multi <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">dms</span><span class="op">$</span><span class="va">ube2i</span>
<span class="co">#&gt; # Deep mutational scanning data</span>
<span class="co">#&gt; # Name: UBE2i</span>
<span class="co">#&gt; # Study: Weile et al. (2017)</span>
<span class="co">#&gt; # Gene: UBE2i</span>
<span class="co">#&gt; # 159 positions</span>
<span class="co">#&gt; # Positional data:</span>
<span class="co">#&gt;    position wt         A       C       D       E       F       G       H      I</span>
<span class="co">#&gt;       &lt;dbl&gt; &lt;chr&gt;  &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">#&gt;  1        1 M     -0.454 -0.356  -0.382  -0.504  -0.434  -0.448  -0.517  -0.487</span>
<span class="co">#&gt;  2        2 S     -0.359  0.175  -0.200  -0.154   0.0421  0.168  -0.0681 -0.362</span>
<span class="co">#&gt;  3        3 G     -0.701 -0.0885  0.0655 -0.124  -0.149   0      -0.310  -0.228</span>
<span class="co">#&gt;  4        4 I     -0.101 -0.127   0.160  -0.0979 -0.365  -0.458  -0.187   0    </span>
<span class="co">#&gt;  5        5 A      0      0.195  -0.588  -0.150  -0.236   0.0219  0.650   0.173</span>
<span class="co">#&gt;  6        6 L     -0.218 -0.151  -0.389  -0.203  -0.205  -0.198  -0.328  -0.535</span>
<span class="co">#&gt;  7        7 S     -0.873 -0.214  -0.545  -0.292  -0.751  -0.385  -0.450  -0.430</span>
<span class="co">#&gt;  8        8 R     -0.214 -0.246  -0.281  -0.0123 -0.267  -0.862  -0.331  -0.989</span>
<span class="co">#&gt;  9        9 L     -0.408 -0.275   0.129  -0.466  -1.15   -0.433  -0.878  -0.551</span>
<span class="co">#&gt; 10       10 A      0     -0.908  -1.12   -0.887  -0.155  -0.361  -0.571  -0.495</span>
<span class="co">#&gt; # … with 149 more rows, and 33 more variables: K &lt;dbl&gt;, L &lt;dbl&gt;, M &lt;dbl&gt;,</span>
<span class="co">#&gt; #   N &lt;dbl&gt;, P &lt;dbl&gt;, Q &lt;dbl&gt;, R &lt;dbl&gt;, S &lt;dbl&gt;, T &lt;dbl&gt;, V &lt;dbl&gt;, W &lt;dbl&gt;,</span>
<span class="co">#&gt; #   Y &lt;dbl&gt;, name &lt;chr&gt;, impute_A &lt;dbl&gt;, impute_C &lt;dbl&gt;, impute_D &lt;dbl&gt;,</span>
<span class="co">#&gt; #   impute_E &lt;dbl&gt;, impute_F &lt;dbl&gt;, impute_G &lt;dbl&gt;, impute_H &lt;dbl&gt;,</span>
<span class="co">#&gt; #   impute_I &lt;dbl&gt;, impute_K &lt;dbl&gt;, impute_L &lt;dbl&gt;, impute_M &lt;dbl&gt;,</span>
<span class="co">#&gt; #   impute_N &lt;dbl&gt;, impute_P &lt;dbl&gt;, impute_Q &lt;dbl&gt;, impute_R &lt;dbl&gt;,</span>
<span class="co">#&gt; #   impute_S &lt;dbl&gt;, impute_T &lt;dbl&gt;, impute_V &lt;dbl&gt;, impute_W &lt;dbl&gt;,</span>
<span class="co">#&gt; #   impute_Y &lt;dbl&gt;</span></code></pre></div>
</div>
<div id="nucleoprotein" class="section level2">
<h2 class="hasAnchor">
<a href="#nucleoprotein" class="anchor"></a>Nucleoprotein</h2>
<p>The data from this study is stored in a custom format, with each row representing a position in the protein and columns for the percentage observed in the post selection library for each amino acid.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df_np</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"doud_2015_np.csv"</span>, package <span class="op">=</span> <span class="st">"deepscanscape"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">" "</span><span class="op">)</span>
<span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="va">df_np</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 498 x 23</span>
<span class="co">#&gt;    POSITION WT    SITE_ENTROPY    PI_A    PI_C    PI_D    PI_E    PI_F    PI_G</span>
<span class="co">#&gt;       &lt;int&gt; &lt;chr&gt;        &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;</span>
<span class="co">#&gt;  1        1 M             3.97 0.0417  0.0411  0.0452  0.0412  0.0398  0.0365 </span>
<span class="co">#&gt;  2        2 A             1.78 0.716   0.00725 0.00540 0.0260  0.00778 0.0118 </span>
<span class="co">#&gt;  3        3 S             3.84 0.118   0.0166  0.00580 0.00620 0.0931  0.0159 </span>
<span class="co">#&gt;  4        4 Q             3.79 0.0261  0.0288  0.00488 0.0387  0.159   0.00851</span>
<span class="co">#&gt;  5        5 G             3.39 0.122   0.0807  0.0117  0.00760 0.00610 0.245  </span>
<span class="co">#&gt;  6        6 T             1.50 0.0450  0.00741 0.00656 0.00604 0.00596 0.00416</span>
<span class="co">#&gt;  7        7 K             1.70 0.00292 0.0116  0.00800 0.0406  0.00620 0.00346</span>
<span class="co">#&gt;  8        8 R             1.90 0.00638 0.0180  0.00794 0.00620 0.00673 0.0173 </span>
<span class="co">#&gt;  9        9 S             3.40 0.121   0.0134  0.00351 0.00706 0.0542  0.00822</span>
<span class="co">#&gt; 10       10 Y             3.87 0.0781  0.0330  0.0414  0.0262  0.0529  0.0112 </span>
<span class="co">#&gt; # … with 488 more rows, and 14 more variables: PI_H &lt;dbl&gt;, PI_I &lt;dbl&gt;,</span>
<span class="co">#&gt; #   PI_K &lt;dbl&gt;, PI_L &lt;dbl&gt;, PI_M &lt;dbl&gt;, PI_N &lt;dbl&gt;, PI_P &lt;dbl&gt;, PI_Q &lt;dbl&gt;,</span>
<span class="co">#&gt; #   PI_R &lt;dbl&gt;, PI_S &lt;dbl&gt;, PI_T &lt;dbl&gt;, PI_V &lt;dbl&gt;, PI_W &lt;dbl&gt;, PI_Y &lt;dbl&gt;</span></code></pre></div>
<p>This format is similar to the “wide” format supported by DeepScanScape, except scores are not yet normalised against the WT value. We will therefore normalise each column by the WT score for that row before using the wide parsing mode and <span class="math inline">\(log_2\)</span> transform to create the standard object.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Set colnames</span>
<span class="va">amino_acids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"C"</span>, <span class="st">"D"</span>, <span class="st">"E"</span>, <span class="st">"F"</span>, <span class="st">"G"</span>, <span class="st">"H"</span>, <span class="st">"I"</span>, <span class="st">"K"</span>, <span class="st">"L"</span>, <span class="st">"M"</span>, <span class="st">"N"</span>, <span class="st">"P"</span>, <span class="st">"Q"</span>, <span class="st">"R"</span>, <span class="st">"S"</span>, <span class="st">"T"</span>, <span class="st">"V"</span>, <span class="st">"W"</span>, <span class="st">"Y"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">df_np</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"position"</span>, <span class="st">"wt"</span>, <span class="st">"entropy"</span>, <span class="va">amino_acids</span><span class="op">)</span>

<span class="co"># Normalise Scores</span>
<span class="va">scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">df_np</span><span class="op">[</span>, <span class="va">amino_acids</span><span class="op">]</span><span class="op">)</span>
<span class="va">wt_score</span> <span class="op">&lt;-</span> <span class="va">scores</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">scores</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="va">df_np</span><span class="op">$</span><span class="va">wt</span>, <span class="va">amino_acids</span><span class="op">)</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">]</span>
<span class="va">scores</span> <span class="op">&lt;-</span> <span class="va">scores</span> <span class="op">/</span> <span class="va">wt_score</span>
<span class="va">df_np</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">df_np</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"position"</span>, <span class="st">"wt"</span><span class="op">)</span><span class="op">]</span>, <span class="va">scores</span><span class="op">)</span>

<span class="co"># Create deep_mutational_scan</span>
<span class="va">dms</span><span class="op">$</span><span class="va">np</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/deep_mutational_scan.html">deep_mutational_scan</a></span><span class="op">(</span><span class="va">df_np</span>, name <span class="op">=</span> <span class="st">"NP"</span>, gene <span class="op">=</span> <span class="st">"NP"</span>, study <span class="op">=</span> <span class="st">"Doud et al. 2015"</span>,
                               scheme <span class="op">=</span> <span class="st">"wide"</span>, trans <span class="op">=</span> <span class="st">"log2"</span>, na_value <span class="op">=</span> <span class="st">"impute"</span><span class="op">)</span>
<span class="va">dms</span><span class="op">$</span><span class="va">np</span>
<span class="co">#&gt; # Deep mutational scanning data</span>
<span class="co">#&gt; # Name: NP</span>
<span class="co">#&gt; # Study: Doud et al. 2015</span>
<span class="co">#&gt; # Gene: NP</span>
<span class="co">#&gt; # 497 positions</span>
<span class="co">#&gt; # Positional data:</span>
<span class="co">#&gt;    position wt          A       C      D       E       F      G      H      I</span>
<span class="co">#&gt;       &lt;dbl&gt; &lt;chr&gt;   &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">#&gt;  1        1 M     -0.500  -0.504  -0.477 -0.503  -0.513  -0.537 -0.424 -0.709</span>
<span class="co">#&gt;  2        2 A      0      -1.28   -1.36  -0.921  -1.26   -1.14  -1.41  -1.40 </span>
<span class="co">#&gt;  3        3 S      0.0350 -0.510  -0.802 -0.784  -0.0312 -0.522 -0.152 -0.241</span>
<span class="co">#&gt;  4        4 Q     -0.0524 -0.0245 -0.518  0.0572  0.451  -0.363 -0.143  0.241</span>
<span class="co">#&gt;  5        5 G     -0.194  -0.309  -0.844 -0.965  -1.03    0     -0.789 -1.02 </span>
<span class="co">#&gt;  6        6 T     -0.799  -1.30   -1.33  -1.36   -1.36   -1.46  -1.14  -0.952</span>
<span class="co">#&gt;  7        7 K     -1.54   -1.16   -1.26  -0.812  -1.33   -1.50  -1.34  -1.06 </span>
<span class="co">#&gt;  8        8 R     -1.31   -1.02   -1.25  -1.32   -1.30   -1.03  -1.37  -1.43 </span>
<span class="co">#&gt;  9        9 S     -0.227  -0.838  -1.21  -1.02   -0.450  -0.973 -1.04  -0.505</span>
<span class="co">#&gt; 10       10 Y     -0.219  -0.458  -0.395 -0.522  -0.327  -0.758 -0.102 -0.179</span>
<span class="co">#&gt; # … with 487 more rows, and 33 more variables: K &lt;dbl&gt;, L &lt;dbl&gt;, M &lt;dbl&gt;,</span>
<span class="co">#&gt; #   N &lt;dbl&gt;, P &lt;dbl&gt;, Q &lt;dbl&gt;, R &lt;dbl&gt;, S &lt;dbl&gt;, T &lt;dbl&gt;, V &lt;dbl&gt;, W &lt;dbl&gt;,</span>
<span class="co">#&gt; #   Y &lt;dbl&gt;, name &lt;chr&gt;, impute_A &lt;dbl&gt;, impute_C &lt;dbl&gt;, impute_D &lt;dbl&gt;,</span>
<span class="co">#&gt; #   impute_E &lt;dbl&gt;, impute_F &lt;dbl&gt;, impute_G &lt;dbl&gt;, impute_H &lt;dbl&gt;,</span>
<span class="co">#&gt; #   impute_I &lt;dbl&gt;, impute_K &lt;dbl&gt;, impute_L &lt;dbl&gt;, impute_M &lt;dbl&gt;,</span>
<span class="co">#&gt; #   impute_N &lt;dbl&gt;, impute_P &lt;dbl&gt;, impute_Q &lt;dbl&gt;, impute_R &lt;dbl&gt;,</span>
<span class="co">#&gt; #   impute_S &lt;dbl&gt;, impute_T &lt;dbl&gt;, impute_V &lt;dbl&gt;, impute_W &lt;dbl&gt;,</span>
<span class="co">#&gt; #   impute_Y &lt;dbl&gt;</span></code></pre></div>
</div>
<div id="hyap65-ww-domain" class="section level2">
<h2 class="hasAnchor">
<a href="#hyap65-ww-domain" class="anchor"></a>hYAP65 WW Domain</h2>
<p>This study associates each variant effect score (the “fitness” column) with a sequence for the entire WW domain - another input format supported by the package. It also contains columns giving the mutations, position and mutant amino acid as comma separated strings as well as additional data from which the fitness was calculated.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df_hyap65</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"araya_2012_hYAP65_ww.tsv"</span>, package <span class="op">=</span> <span class="st">"deepscanscape"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">df_hyap65</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>
<span class="co">#&gt;                              sequence mutation.count       slope    rsquared</span>
<span class="co">#&gt; 1  DVPLPAGWEMAKTRSGQRYFLNHVDQTTKWQDPR              3 -0.04197027 0.001210278</span>
<span class="co">#&gt; 2  DVPLPAGWEMAKTSSDQRYVLNHIDQTTT*QDPR              3  1.32751008 0.858360143</span>
<span class="co">#&gt; 3  EVPLTAGWEMAKTSSGQRYFLNHIDETTTWQDPR              3 -1.07727309 0.733881171</span>
<span class="co">#&gt; 4  DVPLPAGWEKAKTSSGLRYFLNHIDQTTTWQVPR              3  0.56176737 0.195067946</span>
<span class="co">#&gt; 5  YVALPAGWEMAKTSSGQRYFMNHIDQTTTWQDPR              3  1.78624321 0.625087295</span>
<span class="co">#&gt; 6  EVPLPASWEMAKTSSGQRYFLNHIDQTTTWQDPR              2 -0.12632300 0.061684010</span>
<span class="co">#&gt; 7  DVPLPAGWKMDKTSSGQLYFLNHIDQTTTWQDPR              3  0.94473044 0.528514345</span>
<span class="co">#&gt; 8  DVPPPAGWEMAKTCSGQRYFLNHIDQTPTWQDPR              3 -0.88794326 0.720080860</span>
<span class="co">#&gt; 9  DVPLPAGWEMAKTSSGQRYFMNHIDQTTTWLDPR              2 -0.86870768 0.436073420</span>
<span class="co">#&gt; 10 DFQLPAGWEMAKTSSGERYFLNHIDQTTTWQDPR              3  1.29462756 0.958521995</span>
<span class="co">#&gt;      fitness   mutations positions amino.acids</span>
<span class="co">#&gt; 1  0.5547552 23R,33V,38K  23,33,38       R,V,K</span>
<span class="co">#&gt; 2  1.4333626 25D,29V,39*  25,29,39       D,V,*</span>
<span class="co">#&gt; 3  0.2706725 10E,14T,35E  10,14,35       E,T,E</span>
<span class="co">#&gt; 4  0.8430329 19K,26L,41V  19,26,41       K,L,V</span>
<span class="co">#&gt; 5  1.9699196 10Y,12A,30M  10,12,30       Y,A,M</span>
<span class="co">#&gt; 6  0.5232494     10E,16S     10,16         E,S</span>
<span class="co">#&gt; 7  1.0993296 18K,20D,27L  18,20,27       K,D,L</span>
<span class="co">#&gt; 8  0.3086300 13P,23C,37P  13,23,37       P,C,P</span>
<span class="co">#&gt; 9  0.3127726     30M,40L     30,40         M,L</span>
<span class="co">#&gt; 10 1.4010623 11F,12Q,26E  11,12,26       F,Q,E</span></code></pre></div>
<p>We will use the “sequence” parsing mode, in addition to demonstrating the position offset argument, since the variants would otherwise be numbered within the WW domain. This argument requires parsing the input data before passing it to deep_mutational_scan with <code>scheme = NULL</code>. Since the data contains multiple mutations per score scores we will be averaged to generate scores for each single mutant where a single mutant is not available. These averages are not representative when there is an early stop codon, so we will manually filter all rows containing these before parsing. The scores are expressed as enrichment ratios so only require a <span class="math inline">\(log_2\)</span> transformation.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df_hyap65</span> <span class="op">&lt;-</span> <span class="va">df_hyap65</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">"\\*"</span>, <span class="va">df_hyap65</span><span class="op">$</span><span class="va">sequence</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"sequence"</span>, <span class="st">"fitness"</span><span class="op">)</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">df_hyap65</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"sequence"</span>, <span class="st">"score"</span><span class="op">)</span>

<span class="va">parsed_df_hyap65</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parse_deep_scan.html">parse_deep_scan</a></span><span class="op">(</span><span class="va">df_hyap65</span>, scheme <span class="op">=</span> <span class="st">"sequence"</span>, average_multi <span class="op">=</span> <span class="cn">TRUE</span>, position_offset <span class="op">=</span> <span class="fl">169</span><span class="op">)</span>
<span class="va">dms</span><span class="op">$</span><span class="va">hyap65</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/deep_mutational_scan.html">deep_mutational_scan</a></span><span class="op">(</span><span class="va">parsed_df_hyap65</span>, name <span class="op">=</span> <span class="st">"hYAP65"</span>, gene <span class="op">=</span> <span class="st">"hYAP65"</span>, study <span class="op">=</span> <span class="st">"Araya et al. 2012"</span>, 
                                   scheme <span class="op">=</span> <span class="cn">NULL</span>, trans <span class="op">=</span> <span class="st">"log2"</span><span class="op">)</span>
<span class="va">dms</span><span class="op">$</span><span class="va">hyap65</span>
<span class="co">#&gt; # Deep mutational scanning data</span>
<span class="co">#&gt; # Name: hYAP65</span>
<span class="co">#&gt; # Study: Araya et al. 2012</span>
<span class="co">#&gt; # Gene: hYAP65</span>
<span class="co">#&gt; # 34 positions</span>
<span class="co">#&gt; # Positional data:</span>
<span class="co">#&gt;    position wt          A       C       D        E       F        G        H</span>
<span class="co">#&gt;       &lt;dbl&gt; &lt;chr&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;</span>
<span class="co">#&gt;  1      170 D      0.211   0.0358  0       0.160    0.142   0.199    0.0342 </span>
<span class="co">#&gt;  2      171 V      0.112   0.330   0.267   0.144    0.0519  0.293    0.340  </span>
<span class="co">#&gt;  3      172 P      0.0733 -0.226  -0.329  -0.0918  -0.276   0.215   -0.00812</span>
<span class="co">#&gt;  4      173 L      0.653  -0.207  -0.576  -0.00566 -0.234  -0.0431  -0.566  </span>
<span class="co">#&gt;  5      174 P     -0.396  -0.226  -0.329  -0.0311  -0.331  -0.0816  -0.0971 </span>
<span class="co">#&gt;  6      175 A      0      -0.244  -0.118  -0.286   -0.0155 -0.0884  -0.0817 </span>
<span class="co">#&gt;  7      176 G     -0.511  -0.310  -0.595  -0.125   -0.123   0       -0.276  </span>
<span class="co">#&gt;  8      177 W      0.0366 -1.11   -0.506   0.0464  -0.310  -1.02    -0.502  </span>
<span class="co">#&gt;  9      178 E      0.119  -0.150   0.0906  0       -0.176   0.319   -0.590  </span>
<span class="co">#&gt; 10      179 M     -0.0435 -0.243  -0.551   0.341   -0.441  -0.00785 -0.372  </span>
<span class="co">#&gt; # … with 24 more rows, and 34 more variables: I &lt;dbl&gt;, K &lt;dbl&gt;, L &lt;dbl&gt;,</span>
<span class="co">#&gt; #   M &lt;dbl&gt;, N &lt;dbl&gt;, P &lt;dbl&gt;, Q &lt;dbl&gt;, R &lt;dbl&gt;, S &lt;dbl&gt;, T &lt;dbl&gt;, V &lt;dbl&gt;,</span>
<span class="co">#&gt; #   W &lt;dbl&gt;, Y &lt;dbl&gt;, name &lt;chr&gt;, impute_A &lt;dbl&gt;, impute_C &lt;dbl&gt;,</span>
<span class="co">#&gt; #   impute_D &lt;dbl&gt;, impute_E &lt;dbl&gt;, impute_F &lt;dbl&gt;, impute_G &lt;dbl&gt;,</span>
<span class="co">#&gt; #   impute_H &lt;dbl&gt;, impute_I &lt;dbl&gt;, impute_K &lt;dbl&gt;, impute_L &lt;dbl&gt;,</span>
<span class="co">#&gt; #   impute_M &lt;dbl&gt;, impute_N &lt;dbl&gt;, impute_P &lt;dbl&gt;, impute_Q &lt;dbl&gt;,</span>
<span class="co">#&gt; #   impute_R &lt;dbl&gt;, impute_S &lt;dbl&gt;, impute_T &lt;dbl&gt;, impute_V &lt;dbl&gt;,</span>
<span class="co">#&gt; #   impute_W &lt;dbl&gt;, impute_Y &lt;dbl&gt;</span></code></pre></div>
</div>
</div>
<div id="visulalising-and-validating-scans" class="section level1">
<h1 class="hasAnchor">
<a href="#visulalising-and-validating-scans" class="anchor"></a>Visulalising and Validating Scans</h1>
<p>Once the data has been loaded and parsed we can visualise them and identify potentially unusual properties. We will primarily focus on the scan of p53, since that contains data not already incorporated in the landscape. First, we can get an overview of the protein by plotting a heatmap showing all normalised variant effect scores:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_er_heatmap.html">plot_er_heatmap</a></span><span class="op">(</span><span class="va">dms</span><span class="op">$</span><span class="va">p53</span><span class="op">)</span></code></pre></div>
<p><img src="deepscanscape_files/figure-html/heatmap-1.png" width="100%" height="50%" style="display: block; margin: auto;"> We can also examine the distribution of variant effect scores to check that it matches expectations, including through comparison to the overall distribution in the deep landscape. It is expected that scores have a bimodal distribution with a peak around 0 comprising neutral variants and a smaller negative peak of deleterious variants. In this case the new data conforms well to our expectations, suggesting we can meaningfully analyse it using the deep landscape since it contains scores of the same form.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_er_distribution.html">plot_er_distribution</a></span><span class="op">(</span><span class="va">dms</span><span class="op">$</span><span class="va">p53</span><span class="op">)</span></code></pre></div>
<p><img src="deepscanscape_files/figure-html/er_distribution-1.png" width="100%" height="50%" style="display: block; margin: auto;"></p>
<p>In other cases the score distributions can be more unusual and suggest something unexpected about the study. For example, NP has a much larger proportion of deleterious variants than usual, potentially suggesting it is a particularly conserved protein. Similarly the hYAP65 domain doesn’t have a clear deleterious peak, suggesting there are not many particularly problematic variants in this region. Larger deviations, for example a large number of positive scores or a big deviation from the expected peaks can suggest an unusual feature of the experiment or problems with the processing or transformation applied.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span><span class="fu"><a href="../reference/plot_er_distribution.html">plot_er_distribution</a></span><span class="op">(</span><span class="va">dms</span><span class="op">$</span><span class="va">np</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"NP"</span><span class="op">)</span>,
                  <span class="fu"><a href="../reference/plot_er_distribution.html">plot_er_distribution</a></span><span class="op">(</span><span class="va">dms</span><span class="op">$</span><span class="va">hyap65</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"hYAP65 WW Domain"</span><span class="op">)</span>,
                  nrow <span class="op">=</span> <span class="fl">1</span>, common.legend <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="deepscanscape_files/figure-html/er_distribution_unusual-1.png" width="100%" height="50%" style="display: block; margin: auto;"></p>
<p>The package provides some explicit checks to help identify unusual properties. We can demonstrate this by creating a badly transformed version of our UBE2i scan:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bad_scan</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/deep_mutational_scan.html">deep_mutational_scan</a></span><span class="op">(</span><span class="va">df_ube2i</span>, name <span class="op">=</span> <span class="st">"UBE2i"</span>, gene <span class="op">=</span> <span class="st">'UBE2i'</span>, study <span class="op">=</span> <span class="st">'Weile et al. (2017)'</span>,
                                  scheme <span class="op">=</span> <span class="st">"mavedb"</span>, trans <span class="op">=</span> <span class="cn">NULL</span>, na_value <span class="op">=</span> <span class="st">'impute'</span>, duplicates <span class="op">=</span> <span class="st">"mean"</span>,
                                  average_multi <span class="op">=</span> <span class="cn">TRUE</span>, annotate <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="../reference/plot_er_distribution.html">plot_er_distribution</a></span><span class="op">(</span><span class="va">bad_scan</span><span class="op">)</span></code></pre></div>
<p><img src="deepscanscape_files/figure-html/bad_scan-1.png" width="100%" height="50%" style="display: block; margin: auto;"> The <code>check_data</code> function creates a table of potentially unusual properties detected in the dataset. In this case we identify a generally positive score distribution and positions outside the mapped landscape, both of which can suggest an incorrect transformation (as in this case) or unusual score properties.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/check_data.html">check_data</a></span><span class="op">(</span><span class="va">bad_scan</span>, warn <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 1 x 3</span>
<span class="co">#&gt;   id        cause           explanation                                         </span>
<span class="co">#&gt;   &lt;chr&gt;     &lt;chr&gt;           &lt;chr&gt;                                               </span>
<span class="co">#&gt; 1 positive… Median ER scor… Fitness is potentially inverted either by experimen…</span></code></pre></div>
</div>
<div id="analysis-using-the-deep-landscape" class="section level1">
<h1 class="hasAnchor">
<a href="#analysis-using-the-deep-landscape" class="anchor"></a>Analysis using the Deep Landscape</h1>
<p>Once a standardised deep scan object has been created and checked we can annotate it with information from the deep mutational landscape. This projects the new positions onto the PCA and UMAP space created from the overall landscape and assigns each position to an amino acid subtype based on the nearest cluster centroid. The <code>describe_clusters</code> function can be used to give information about each position based on the assigned subtype.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">annotated_dms</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annotate.html">annotate</a></span><span class="op">(</span><span class="va">dms</span><span class="op">$</span><span class="va">p53</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/describe_clusters.html">describe_clusters</a></span><span class="op">(</span><span class="va">annotated_dms</span><span class="op">)</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 10 x 8</span>
<span class="co">#&gt;    name  position wt    cluster global_cluster_f… group description    notes    </span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;          &lt;chr&gt;    </span>
<span class="co">#&gt;  1 p53          1 T     T5                0.0615  Nega… Weak selectio… Weak sel…</span>
<span class="co">#&gt;  2 p53          2 Y     Y3                0.227   Arom… Generally sel… Generall…</span>
<span class="co">#&gt;  3 p53          3 Q     Q2                0.2     Not … Anything but … Anything…</span>
<span class="co">#&gt;  4 p53          4 G     G6                0.0512  Uniq… Largely permi… Largely …</span>
<span class="co">#&gt;  5 p53          5 S     S1                0.398   Smal… Only tolerate… Only tol…</span>
<span class="co">#&gt;  6 p53          6 Y     Y3                0.227   Arom… Generally sel… Generall…</span>
<span class="co">#&gt;  7 p53          7 G     G6                0.0512  Uniq… Largely permi… Largely …</span>
<span class="co">#&gt;  8 p53          8 F     FO                0.00722 Outl… Outlier        &lt;NA&gt;     </span>
<span class="co">#&gt;  9 p53          9 R     R1                0.461   Posi… Intolerant of… Intolera…</span>
<span class="co">#&gt; 10 p53         10 L     LO                0.0561  Outl… Outlier        &lt;NA&gt;</span></code></pre></div>
<p>We can also visualise the new positions on the annotated deep landscape, for example against the distribution of SIFT4G scores or FoldX <span class="math inline">\(\Delta\Delta G\)</span> predictions. This plot also allows us to highlight specific positions of interest:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span><span class="fu"><a href="../reference/plot_landscape.html">plot_landscape</a></span><span class="op">(</span><span class="va">annotated_dms</span>, feature <span class="op">=</span> <span class="st">"mean_sift"</span><span class="op">)</span>,
                  <span class="fu"><a href="../reference/plot_landscape.html">plot_landscape</a></span><span class="op">(</span><span class="va">annotated_dms</span>, feature <span class="op">=</span> <span class="st">"total_energy"</span>, highlight <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">10</span>, <span class="fl">20</span><span class="op">)</span><span class="op">)</span>,
                  nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p><img src="deepscanscape_files/figure-html/deep_landscape-1.png" width="100%" height="50%" style="display: block; margin: auto;"></p>
<p>These plots can be made for any numerical features of the deep landscape dataset and two special modes, projecting landscape position density and the most prevalent subtype group at that position:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span><span class="fu"><a href="../reference/plot_landscape.html">plot_landscape</a></span><span class="op">(</span><span class="va">annotated_dms</span>, feature <span class="op">=</span> <span class="st">"count"</span><span class="op">)</span>,
                  <span class="fu"><a href="../reference/plot_landscape.html">plot_landscape</a></span><span class="op">(</span><span class="va">annotated_dms</span>, feature <span class="op">=</span> <span class="st">"cluster"</span><span class="op">)</span>,
                  nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p><img src="deepscanscape_files/figure-html/deep_landscape_modes-1.png" width="100%" height="50%" style="display: block; margin: auto;"></p>
<p>Finally we can visualise the frequencies of each subtype in the new dataset, with an option to explicitly compare them to the main landscape:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_cluster_frequencies.html">plot_cluster_frequencies</a></span><span class="op">(</span><span class="va">annotated_dms</span><span class="op">)</span></code></pre></div>
<p><img src="deepscanscape_files/figure-html/subtype_freqs-1.png" width="100%" height="50%" style="display: block; margin: auto;"></p>
</div>
<div id="combining-deep-scans" class="section level1">
<h1 class="hasAnchor">
<a href="#combining-deep-scans" class="anchor"></a>Combining Deep Scans</h1>
<p>In some cases you may want to analyse multiple scans together, for example if you have several scans of one gene under different conditions or several similar genes in the same condition. The deep_mutational_scan object supports combined data of this sort, allowing the same analyses. The first stage is combining data from multiple scans into one object, which can be visualised in a similar manner:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">combined_scans</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bind_scans.html">bind_scans</a></span><span class="op">(</span><span class="va">dms</span>, annotate_missing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span><span class="fu"><a href="../reference/plot_er_heatmap.html">plot_er_heatmap</a></span><span class="op">(</span><span class="va">combined_scans</span><span class="op">)</span>, <span class="fu"><a href="../reference/plot_er_distribution.html">plot_er_distribution</a></span><span class="op">(</span><span class="va">combined_scans</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p><img src="deepscanscape_files/figure-html/binding-1.png" width="100%" height="80%" style="display: block; margin: auto;"></p>
<p>Similar deep landscape analyses can also be performed, for example projecting points onto the deep landscape:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_landscape.html">plot_landscape</a></span><span class="op">(</span><span class="va">combined_scans</span>, feature <span class="op">=</span> <span class="st">"cluster"</span><span class="op">)</span></code></pre></div>
<p><img src="deepscanscape_files/figure-html/combined_deep_landscape-1.png" width="100%" height="50%" style="display: block; margin: auto;"></p>
</div>
<div id="re-clustering-the-deep-landscape" class="section level1">
<h1 class="hasAnchor">
<a href="#re-clustering-the-deep-landscape" class="anchor"></a>Re-clustering the Deep Landscape</h1>
<p>Finally, the package provides a function to apply the amino acid subtype clustering procedure to new data, including in combination with the original dataset. This is only likely to be useful when a reasonably large number of additional studies are added and the downstream analysis is more technical so it is only recommended for more advanced users. Some basic parameters are provided to tweak the clustering procedure but larger changes will require modifying the code and so are also more complicated. A basic example of re-clustering is shown below, using our example p53 scan that was not already in the dataset:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">reclust</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/recluster.html">recluster</a></span><span class="op">(</span><span class="va">dms</span><span class="op">$</span><span class="va">p53</span>, add_combined <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="../reference/plot_recluster.html">plot_recluster</a></span><span class="op">(</span><span class="va">reclust</span><span class="op">)</span></code></pre></div>
<p><img src="deepscanscape_files/figure-html/recluster-1.png" width="100%" height="50%" style="display: block; margin: auto;"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Alistair Dunham.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
